iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +548,638 @@ fn close_button() -> Node<Msg> {
+        assert_eq!(&pair, &(Select::CommandJob(77, vec![162, 163]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(309));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162, 163], vec![309]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(308));
+        assert_eq!(
+            &pair,
+            &(Select::CommandJobSteps(77, vec![162, 163], vec![308, 309]), true)
+        );
+
+        let pair = interpret_click(&pair.0, TypedId::Job(163));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162], vec![308, 309]), false));
+
+        let pair = interpret_click(&pair.0, TypedId::Job(162));
+        assert_eq!(&pair, &(Select::Command(77), false));
+    }
+
+    /// There could the the problem, when `schedule_fetch_tree` made the api requests
+    /// in one order, but during delays or something the responses may come in different
+    /// order. In all such cases, however, model should remain consistent,
+    /// and the test checks this.
+    #[test]
+    fn test_async_handlers_consistency() {
+        let mut rng = Xoroshiro64Star::seed_from_u64(485369);
+        let db = build_db();
+        let mut model = Model::default();
+        let cmd_ids = db.all_cmds.iter().map(|x| x.id).collect::<Vec<_>>();
+        let selects = generate_random_selects(&db, &mut rng, 100);
+        for select in selects {
+            let permutations = vec![[1, 2, 3], [1, 3, 2], [2, 1, 3], [2, 3, 1], [3, 1, 2], [3, 2, 1]];
+            for permutation in permutations {
+                let cmd_id = cmd_ids[rng.next_u32() as usize % cmd_ids.len()];
+                let (c, j, s) = prepare_subset(&db, cmd_id);
+                model.clear();
+                model.select = select.clone();
+                model.assign_commands(&c);
+                model.assign_jobs(&j);
+                model.assign_steps(&s);
+                let expected_cmd = model.commands_view.clone();
+                let expected_jobs = model.jobs_graph.clone();
+                let expected_steps = model.steps_view.clone();
+
+                model.clear();
+                model.select = select.clone();
+                // we simulate, that FetchCommands, FetchJobs and FetchSteps come in arbitrary order
+                for p in &permutation {
+                    match p {
+                        1 => model.assign_commands(&c),
+                        2 => model.assign_jobs(&j),
+                        3 => model.assign_steps(&s),
+                        _ => unreachable!(),
+                    }
+                }
+                // compare debug strings since there is no Eq instance for iml_wire_types
+                assert_eq!(format!("{:?}", model.commands_view), format!("{:?}", expected_cmd));
+                assert_eq!(format!("{:?}", model.jobs_graph), format!("{:?}", expected_jobs));
+                assert_eq!(format!("{:?}", model.steps_view), format!("{:?}", expected_steps));
```

Comment:
Now I see that you compare `commands_view`, not `commands` :)

But I still think this test has no value, because you call `refresh_view()` and `check_consistency()` on each `assign_foo()` invocation, and the last one will have all the data needed to refresh the view.

It is still unclear to me, so here are the questions:

1. How do you define the consistent state?
2. If you define a consistent state, why does it depend on how we get to the state?
3. Can you provide with an example when this test fails?

What I tried:
1. Remove `match select` completely from `check_consistency()`  - the test is passing
2. Use different seeds for the RNG - the test is passing
3. Replace some `true` to `false` in  `check_consistency()` - the test is still passing (!)
4. Combinations of 2 and 3.

So what exactly is this test for? :)

Reply:
Hm, I changed the piece to
```
                if cmd0.is_some() {
                    ls[0] = false;
                }
```
and with the seed `23` these 100 samples was enough `¯\_(ツ)_/¯` to falsify the test:
![image](https://user-images.githubusercontent.com/940363/80349414-0a242d80-8891-11ea-8785-24633c98690c.png)

Reply:
Ok, let me describe what I am doing and what problem I'm trying to solve.

The user (implicitly, via GUI) sets the selection to some value, and this makes a number of requests to send to the API. Assume, the requests caused to grab some commands, jobs and steps.

Now what if the steps come first? Should we discard them or write them into the model?
1. if we discard them, then we need to implement some logic to resend the request again to finally get the batch.
2. if we write it to the model, the model now potentially have old commands, no jobs, and the fresh steps.

Option 1 is dangerous, because the logic, that makes the request again can cause `fire loop`, i.e. in one handler we perform the check, and and fire the request, and in another handler we perform slightly different check and fire the different request, and this two handlers makes firing to each other indefinitely. It needs some work to prove that the process converges, which I don't want to do.

Option 2 is better, however to ensure the model is always consistent, I do the following:
- when a message is received, it unconditionally written to a `raw` part of the state
- there is a logic, that rebuilds the `view` part of the state, and ensures that the `view` is always consistent
- the rest of the logic depends only on the `view` part.

In this case, raw part of the model is:
```rust
    pub commands: HashMap<u32, Arc<RichCommand>>,
    pub jobs: HashMap<u32, Arc<RichJob>>,
    pub steps: HashMap<u32, Arc<RichStep>>,
```
and the view part of the model is
```
    pub commands_view: Vec<Arc<RichCommand>>,
    pub jobs_graph: JobsGraph,
    pub steps_view: Vec<Arc<RichStep>>,
```

As a side note, this problem is similar to the implementation of distributed queues, when there is an ordering between the messages and at some point we receive a message out of order, we maintain a local buffer to wait for some time for the needed message while the clients just see the messages always in order and can rely on this.

Reply:
> if cmd0.is_some() {
>                   ls[0] = false;
>                }

But this is the opposite to the above code (initial ls[0] value). Not surprising it fails :)